### PR TITLE
[5.2] QueuingConfiguration for jobs

### PR DESCRIPTION
**Pull request by halaei**
With added functionality for changing connection.
A proposal to consider for [this issue](https://github.com/laravel/framework/issues/9896): (DispatchesJobs cannot queue to a different connection in Laravel 5.1)

### CODE REVIEW

File: src/Illuminate/Bus/BusServiceProvider.php [hunk 1]
```diff
@@ -21,7 +21,11 @@ class BusServiceProvider extends ServiceProvider
     public function register()
     {
         $this->app->singleton('Illuminate\Bus\Dispatcher', function ($app) {
-            return new Dispatcher($app, function () use ($app) {
+            return new Dispatcher($app, function ($command) use ($app) {
+                if (isset($command->queue) && $command->queue instanceof QueuingConfiguration) {
+                    return \Queue::connection($command->queue->connection);
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/BusServiceProvider.php]: no way

File: src/Illuminate/Bus/Dispatcher.php [hunk 2]
```diff
@@ -250,20 +246,24 @@ public function dispatchToQueue($command)
      *
      * @param  \Illuminate\Contracts\Queue\Queue  $queue
      * @param  mixed  $command
-     * @return void
+     * @return mixed
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/Dispatcher.php]: no

File: src/Illuminate/Bus/Dispatcher.php [hunk 3]
```diff
@@ -250,20 +246,24 @@ public function dispatchToQueue($command)
...
      */
     protected function pushCommandToQueue($queue, $command)
     {
-        if (isset($command->queue) && isset($command->delay)) {
-            return $queue->laterOn($command->queue, $command->delay, $command);
-        }
+        if (isset($command->queue) && $command->queue instanceof QueuingConfiguration) {
+            /** @var QueuingConfiguration $config */
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/Dispatcher.php]: no

File: src/Illuminate/Bus/Dispatcher.php [hunk 4]
```diff
@@ -250,20 +246,24 @@ public function dispatchToQueue($command)
...
+            $config = $command->queue;
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/Dispatcher.php]: cs

File: tests/Bus/BusDispatcherTest.php [hunk 5]
```diff
@@ -167,6 +168,13 @@ public function queue($queue, $command)
 
 class BusDispatcherTestSpecificQueueAndDelayCommand implements Illuminate\Contracts\Queue\ShouldQueue
 {
-    public $queue = 'foo';
-    public $delay = 10;
+    /**
+     * @var Illuminate\Bus\QueuingConfiguration
```
Comment (GrahamCampbell) [inline tests/Bus/BusDispatcherTest.php]: no

Comment (GrahamCampbell) [inline tests/Bus/BusDispatcherTest.php]: don't need this in a test stub - remove it please

**Comment by halaei** [comment]
Minor issues aside, is the idea generally right?

**Comment by taylorotwell** [comment]
This feels like a lot of code when you could just do:

`Queue::connection('foo')->push(new Job);

**Comment by halaei** [comment]
Then, that may hold for `ShouldQueue`, `QueueingDispatcher` interfaces and `Queueable` classes and a portion of Dispatcher.

The Bus is partially handling queue management, why not do that more thorough?

And I guess there is a shortage of documentations about queues. Could there be something about QueueManager and Queue contract?

**Pull request closed without merging**